core/math/fft.h

```diff
@@ -32,15 +32,29 @@ namespace Math {
 class FFT1D {
 public:
   FFT1D(size_t N, int direction) : _data(N), direction(direction) {
-    fftw_complex *p = reinterpret_cast<fftw_complex *>(&(_data[0]));
+    fftw_complex *p = reinterpret_cast<fftw_complex *>(_data.data());
     _plan = fftw_plan_dft_1d(N, p, p, direction, FFTW_MEASURE);
   }
 
   FFT1D(const FFT1D &other) : _data(other._data.size()), direction(other.direction) {
-    fftw_complex *p = reinterpret_cast<fftw_complex *>(&(_data[0]));
-    _plan = fftw_plan_dft_1d(other._data.size(), p, p, direction, FFTW_MEASURE);
+    fftw_complex *p = reinterpret_cast<fftw_complex *>(_data.data());
+    _plan = fftw_plan_dft_1d(_data.size(), p, p, direction, FFTW_MEASURE);
   }
 
+  FFT1D &operator=(const FFT1D &other) {
+    fftw_destroy_plan(_plan);
+    _data.resize(other._data.size());
+    direction = other.direction;
+    fftw_complex *p = reinterpret_cast<fftw_complex *>(_data.data());
```

Comment:
Not sure how to get rid of the need for an explicit cast here... `fftw_complex` is basically a `double[2]`, and we're mapping it directly onto a `std::complex<double>`. Probably fragile if for any reason FFTW or the C++ standard decides to change how they represent their complex types, but I don't see that happening any time soon. The core guidelines suggest you could get about it using a `std::variant`, but it just feels like overkill. Are we happy to leave this as-is?

Reply:
I agree that in this case `reinterpret_cast` makes sense since we are interfacing with a C library. One alternative solution is suggested here https://stackoverflow.com/a/4214426 and consists in just using `memcpy` as pretty much all compilers will optimise the memcpy out.

Reply:
memcpy is not an option, as what we're trying to do is use the std::vector<cdouble> as the backing store for the FFT call, relying on the fact that they have exactly the same memory layout. The solution in that post would only optimise out the memcpy in the static case, where it can clearly tell that the buffer being copied is not used thereafter, and has a compile-time known size. Not the case for us here. 

The better solution would be to use that union concept (similar to `std::variant`, I guess). But it's still overkill IMO. To be honest, I don't mind that warning remaining a thing, since there is a risk of problems if the representations for `fftw_complex` or `std::complex<double>` change for some reason. We could just put a comment to that effect, as a reminder that we are making an assumption here.